src/quantuminspire/projectq/backend_qx.py

```diff
@@ -743,13 +738,15 @@ def __init__(self, qubit_id: int) -> None:
     def _get_measured_qubit_iterator(self, measurement_block_index: int) -> Iterator[int]:
         """
         Get an iterator for the measured qubits. The iterator returns the indexes of the qubits that are measured.
+        With fsp, the simulator will return all qubits as measured, but when there were measurements in the algorithm
+        we want to get these measured qubits only.
 
         :param measurement_block_index: measurement block index for multi measurement results.
 
         :return:
             Iterator which iterated the indexes of the measured qubits
         """
-        if self._full_state_projection:
+        if self._full_state_projection and len(self._measured_ids) != 0:
```

Comment:
Not exactly. `_store` switches to non-fsp only when gates are found after measurements.
For example the user send in the algorithm:
```
H | q1
CNOT | (q1, q4)
Measure | q1
```
Then, because we explicitly asked to only measure q1, we want in the resulting iterator only q1. The situation is that SDK sends the algorithm in as:
```
H | q1
CNOT | (q1, q4)
```
(without the measurement to run it as fsp, because with the measurement it is non-deterministic) and post-process the result based upon the measurements that were 'deleted'

When the user sent in the algorithm without measurements:
```
H | q1
CNOT | (q1, q4)
```
we post-process it as all qubits are measured (The measurement mask from the result is used in this case).